core/domain/stats_domain.py

```diff
@@ -466,30 +471,116 @@ def validate(self):
+
+
+class AnswerFrequencyList(object):
+    """Domain object that represents a list of AnswerOccurrences."""
+    def __init__(self, answer_occurrences=None):
+        """Initialize domain object for answer frequency list for a given list
+        of AnswerOccurrence objects (default is empty list).
+        """
+        self.answer_occurrences = (
+            answer_occurrences if answer_occurrences else [])
+
+    def add_answer(self, answer_occurrence):
+        """Adds a new AnswerOccurrence object."""
+        self.answer_occurrences.append(answer_occurrence)
+
+    def to_raw_type(self):
+        return [
+            answer_occurrence.to_raw_type()
+            for answer_occurrence in self.answer_occurrences]
+
+    @classmethod
+    def from_raw_type(cls, answer_occurrence_list):
+        return cls([
+            AnswerOccurrence.from_raw_type(answer_occurrence_dict)
+            for answer_occurrence_dict in answer_occurrence_list])
+
+
+class CategorizedAnswerFrequencyLists(object):
+    """AnswerFrequencyLists that are categorized based on arbitrary categories.
+    """
+    def __init__(self, categorized_answer_freq_lists=None):
+        """Initialize domain object for categorized answer frequency lists for
+        a given dict (default is empty).
+        """
+        self.categorized_answer_freq_lists = (
+            categorized_answer_freq_lists
+            if categorized_answer_freq_lists else {})
+
+    def to_raw_type(self):
+        return {
+            category: answer_frequency_list.to_raw_type()
+            for category, answer_frequency_list in (
+                self.categorized_answer_freq_lists.iteritems())
+        }
+
+    @classmethod
+    def from_raw_type(cls, categorized_frequency_dict):
+        return cls({
+            category: AnswerFrequencyList.from_raw_type(answer_occurrence_list)
+            for category, answer_occurrence_list in (
+                categorized_frequency_dict.iteritems())
+        })
+
+
 class StateAnswersCalcOutput(object):
     """Domain object that represents output of calculations operating on
     state answers.
     """
-
     def __init__(self, exploration_id, exploration_version, state_name,
-                 calculation_id, calculation_output):
+                 calculation_id, calculation_output_type, calculation_output):
         """Initialize domain object for state answers calculation output.
 
-        calculation_output is a list of dicts containing the results of the
-        specific calculation.
+        calculation_output one of AnswerFrequencyList or
+            CategorizedAnswerFrequencyLists.
         """
         self.exploration_id = exploration_id
         self.exploration_version = exploration_version
         self.state_name = state_name
         self.calculation_id = calculation_id
+        self.calculation_output_type = calculation_output_type
         self.calculation_output = calculation_output
+        if calculation_output_type:
```

Comment:
Is there a case where this doesn't exist?

Reply:
Handled at a higher level.

Yes, if existing aggregation data is loaded by this code. We should clear StateAnswersCalcOutputModel in prod to work around this. If this ever happens with new data it will fail validation (as expected).

Reply:
Ah OK. We'll need some instructions for the release coordinator then, I'm guessing...

Reply:
(and a test plan)

Reply:
Noted this in the weekly tech lead release notes.